core/include/vecmem/containers/impl/static_array.ipp

```diff
@@ -122,20 +122,141 @@ static_array<T, N>::data(void) const {
     /*
      * Return a pointer to the underlying data, but the elements are const.
      */
-    return v;
+    return m_array;
 }
 
 template <typename T, std::size_t N>
-template <typename std::size_t... Is, typename... Tp>
-VECMEM_HOST_AND_DEVICE static_array<T, N>::static_array(
-    std::index_sequence<Is...>, Tp&&... a) {
-    /*
-     * Construct the array from an enumerated list of elements of arbitrary
-     * size. This is a fold expression which iterates over the elements of
-     * Is (the indices) and Tp (the values), and inserts them into the inner
-     * array using a comma expression.
-     */
-    (static_cast<void>(v[Is] = a), ...);
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::iterator
+static_array<T, N>::begin() {
+
+    return m_array;
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::const_iterator
+static_array<T, N>::begin() const {
+
+    return m_array;
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::const_iterator
+static_array<T, N>::cbegin() const {
+
+    return begin();
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::iterator
+static_array<T, N>::end() {
+
+    return m_array + N;
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::const_iterator
+static_array<T, N>::end() const {
+
+    return m_array + N;
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::const_iterator
+static_array<T, N>::cend() const {
+
+    return end();
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::reverse_iterator
+static_array<T, N>::rbegin() {
+
+    return reverse_iterator(end());
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr
+    typename static_array<T, N>::const_reverse_iterator
+    static_array<T, N>::rbegin() const {
+
+    return const_reverse_iterator(end());
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr
+    typename static_array<T, N>::const_reverse_iterator
+    static_array<T, N>::crbegin() const {
+
+    return rbegin();
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr typename static_array<T, N>::reverse_iterator
+static_array<T, N>::rend() {
+
+    return reverse_iterator(begin());
+}
+
+template <typename T, std::size_t N>
+VECMEM_HOST_AND_DEVICE constexpr
+    typename static_array<T, N>::const_reverse_iterator
+    static_array<T, N>::rend() const {
```

Comment:
I just noticed this now... We'll definitely need to tweak `clang-format` a little further... :stuck_out_tongue:

Reply:
Sure. In all honesty though, this is an absolute beast of a function signature to format. Just changing the return type to `auto` probably already helps a lot.

Reply:
Or even going with 

```cpp
template <typename T, std::size_t N>
VECMEM_HOST_AND_DEVICE constexpr auto
    static_array<T, N>::rend() const -> const_reverse_iterator {
```